### DISCUSSION TITLE
Undo/Redo Approaches

### DISCUSSION THREAD

**Discussion by wyhinton** [fltk-rs/fltk-rs#624] [General]
Hello all,

I've been grappling with the issue of how to implement undo/redo for a complex app for a while now. I cant seem to get any of the popular undo/redo crates working in my basic app ([rundo](https://github.com/M-Adoo/rundo), [redo](https://crates.io/crates/redo), or [undo](https://crates.io/crates/undo). Lets say you have an app where you want to be able to add and image frame to a gallery, delete it, and be able to redo or undo those actions. You're using a container and RefCell to mutate your apps state, and each time you mutate you want to generat some kind of history/event log that can than be undone. How might you do that?

**Comment by MoAlyousef** [comment]
Looking at the undo & redo crates, states seem to be stored in a Vec.
You can also check this example here:
https://github.com/tdryer/7guis-fltk-rs/blob/master/src/circles.rs
Which is an implementation of the circle drawing 7guis task with support for undo/redo. States are manually stored in a Vec, while keeping the current index of the vec as the current state, undoing decrements the index while redoing increments it, both fetching the associated relevant data (which can be image paths and dimensions in your app).

**Reply by wyhinton** [reply]
Ah i see! Do you think the approach from that 7guis example scales well? What if your AppState looks more like this:
```
pub struct AppState{
    pub offset: Vector2<f32>,
    pub images: Vec<Box<dyn CanvasImageExt>>,
    pub buttons: Vec<Box<dyn ToolBarButtonExt>>,
    pub masks: Vec<CanvasItem>,
    pub screen_pos: Vector2<i32>,
    pub world_pos: Vector2<f32>,
    pub drag_coord: Vector2<i32>,
    pub drag_start: Vector2<i32>,
    pub zoom_scale: Vector2<f32>,
    pub selection_box: CanvasItem,
    pub sender: fltk::app::Sender<Message>,
    pub active_tool: ToolType,
    pub clipping: bool, 
    pub scroll_bar: ScrollBar,
    pub canvas_size: (f32, f32),
    pub canvas: CanvasItem,
    pub selection_path: Vec<Vector2<f32>>,
    pub active_image: usize,
    pub selected_mask: Option<usize>,
    pub selected_button: Option<Box<dyn ToolBarButtonExt>>,
    pub active_resize_handle: Option<DragHandle>,
    pub cursor: fltk::enums::Cursor,
}
```
When you have a complex app state, do you think a more functional approach could make more sense? In a more functional approach, I'm imaging that all the actions used to mutate the app state would generate history data. This history data can then be traversed to undo those mutation. Instead of storing many copies of the entire app state, you just store the changes. Maybe it would look something like this?
```
impl AppContainer{
    fn new(inner: AppState) -> AppContainer{
        let app = AppState{
            gallery: inner.gallery.clone(),
            // cards: inner.cards.clone(),
            sender: inner.sender.clone(),
            
        };
        return AppContainer(Rc::new(RefCell::new(app)));
    }
    fn delete_card(&self, id: Uuid)->Result<History)> {
        (self.0.borrow_mut()).gallery.cards.retain(|x| x.id != id)
    }
    fn add_card(&self, path: String)->Result<History>{
        let s = (self.0.borrow_mut()).sender.clone();
        let new_card = Card::new(&path, s.clone()).unwrap();
        (self.0.borrow_mut()).gallery.cards.push(new_card);
    }
}
```
I think the undo crate takes this approach, can't figure out how to get it working with a RefCell wrapped AppState though...

**Reply by MoAlyousef** [reply]
I honestly don't know what the best approach would be in this case. But yeah I agree that storing multiple copies of a complex AppState would be inefficient. I think trying to isolate certain actions would be better. 

You can bypass having to wrap your AppState in a RefCell by storing only certain member values in a RefCell (those that are passed into capturing callbacks for example), or you can break down the AppState into smaller structs but you'll have to pass them to most places you pass your AppState. Or you can use lazy_static to store a Mutex<Vec<AppState>> which would have interior mutability, is safe, light to copy, doesn't require passing all around and is only instantiated once.

**Reply by wyhinton** [reply]
Thanks for you thoughts, yes I know its a rather complex issue, I feel like I haven't really been able to find any examples of  undo/redo for a complex rust app yet...I think I will try the lazy_static Mutex<Vec> approach, maybe its a little hacky but I think it should get the job done.

**Reply by wyhinton** [reply]
Some asking around on reddit and stack overflow got me some nice responses and possible undo/redo systems:

- See [this ](https://stackoverflow.com/questions/67238095/the-trait-cannot-be-made-into-an-object-because-method-merge-references-the-s)answer on stackoverflow which using the undo crate
